File: src/main/java/com/shelterhub/controller/MedicalRecordController.java
package com.shelterhub.controller;

import com.shelterhub.dto.request.MedicalRecordRequest;
import com.shelterhub.dto.response.MedicalRecordResponse;
import com.shelterhub.service.MedicalRecordService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping(value = "/v1/medical-record", produces = MediaType.APPLICATION_JSON_VALUE)
public class MedicalRecordController {
    @Autowired
    private MedicalRecordService medicalRecordService;

    @GetMapping
    public ResponseEntity<List<MedicalRecordResponse>> getAll() {
        var medicalRecords = medicalRecordService.getAll();
        if (medicalRecords.isEmpty())
            return ResponseEntity.notFound()
                    .build();

        return ResponseEntity.ok(medicalRecords);
    }

    @GetMapping("/{id}")
    public ResponseEntity<MedicalRecordResponse> getById(@PathVariable UUID id) {
        var medicalRecordOptional = medicalRecordService.getById(id);
        return ResponseEntity
                .of(medicalRecordOptional);
    }

    @PostMapping
    public ResponseEntity<MedicalRecordResponse> create(@RequestBody MedicalRecordRequest medicalRecord) {
        var medicalResponse = medicalRecordService.create(medicalRecord);

        return ResponseEntity.ofNullable(medicalResponse);
    }

    @PutMapping("/{id}")
    public ResponseEntity<MedicalRecordResponse> update(
            @PathVariable UUID id,
            @RequestBody MedicalRecordRequest medicalRecord
    ) {
        var medicalRecordPersisted = medicalRecordService.update(medicalRecord, id);

        return ResponseEntity.ofNullable(medicalRecordPersisted);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity delete(@PathVariable UUID id) {
        var medicalRecord = medicalRecordService.deleteById(id);
        if (medicalRecord.isPresent())
            return ResponseEntity
                    .noContent()
                    .build();

        return ResponseEntity
                .notFound()
                .build();
    }
}

File: src/main/kotlin/br/com/shelterhubmanagementapi/controller/MedicalRecordController.kt
package br.com.shelterhubmanagementapi.controller

import br.com.shelterhubmanagementapi.dto.request.MedicalRecordRequest
import br.com.shelterhubmanagementapi.dto.response.MedicalRecordResponse
import br.com.shelterhubmanagementapi.service.MedicalRecordService
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.UUID

@RestController
@RequestMapping(value = ["/v1/medical-record"], produces = [MediaType.APPLICATION_JSON_VALUE])
class MedicalRecordController(private val medicalRecordService: MedicalRecordService) {
    @GetMapping
    suspend fun getAll(): ResponseEntity<List<MedicalRecordResponse>> {
        val medicalRecords = medicalRecordService.getAll()
        return if (medicalRecords.isEmpty()) {
            ResponseEntity.notFound()
                .build()
        } else {
            ResponseEntity.ok(
                medicalRecords,
            )
        }
    }

    @GetMapping("/{id}")
    suspend fun getById(
        @PathVariable id: UUID,
    ): ResponseEntity<MedicalRecordResponse> {
        val medicalRecordOptional = medicalRecordService.getById(id)
        return ResponseEntity
            .of(medicalRecordOptional)
    }

    @PostMapping
    suspend fun create(
        @RequestBody medicalRecord: MedicalRecordRequest,
    ): ResponseEntity<MedicalRecordResponse> {
        val medicalResponse = medicalRecordService.create(medicalRecord)
        return ResponseEntity.ofNullable(medicalResponse)
    }

    @PutMapping("/{id}")
    suspend fun update(
        @PathVariable id: UUID,
        @RequestBody medicalRecord: MedicalRecordRequest,
    ): ResponseEntity<MedicalRecordResponse> {
        val medicalRecordPersisted = medicalRecordService.update(medicalRecord, id)
        return ResponseEntity.ofNullable(medicalRecordPersisted)
    }

    @DeleteMapping("/{id}")
    suspend fun delete(
        @PathVariable id: UUID?,
    ): ResponseEntity<*> {
        val medicalRecord = medicalRecordService.deleteById(id)
        return if (medicalRecord.isPresent) {
            ResponseEntity.noContent()
                .build<Unit>()
        } else {
            ResponseEntity.notFound()
                .build<Unit>()
        }
    }
}
